### ISSUE TITLE
Re-evaluate second (length) index

### ISSUE THREAD

**Issue by k0ekk0ek** [NLnetLabs/simdzone#30]
Unlike json, fields (strings, numbers, etc) in zone files are not cleanly terminated when it comes to character strings made up by contiguous non-whitespace, non-special characters.

Finding the end for a quoted section is "relatively" easy and the same algorithm used in simdjson can be used. i.e. load a block of bytes, write it out and try to locate the terminating quote (`"`) and escape sequence (`\\`), increase the length by the terminating quote (possibly handle an escape sequence too) and repeat the process until a terminating quote is located.

Finding the end for a contiguous section is harder as we need to locate both characters in the space and special classes.

```
// indexed by low nibbles
const uint8_t contiguous_table[16] = {
  0x00, // 0x00 : "\0" : 0x00 -- end-of-file, ends record/comment
        //         " " : 0x20 -- space (clash)
  0x00, // 0x01
  0x22, // 0x02 : "\"" : 0x22 -- start/end quoted
  0x00, // 0x03
  0x00, // 0x04
  0x00, // 0x05
  0x00, // 0x06
  0x00, // 0x07
  0x28, // 0x08 :  "(" : 0x28 -- starts grouped
  0x09, // 0x09 : "\t" : 0x09 -- space
        //         ")" : 0x29 -- ends grouped (clash)
  0x0a, // 0x0a : "\n" : 0x0a -- end-of-line, ends record/comment
  0x3b, // 0x0b :  ";" : 0x3b -- start comment
  0x00, // 0x0c : "\\" : 0x5c -- start escape sequence
        //                       (does not delimit, separate lookup)
  0x0d, // 0x0d : "\r" : 0x0d -- space
  0x00, // 0x0e
  0x00, // 0x0f
};
```

While `_mm_shuffle_epi8` can be used to determine which byte to compare for each byte, that process becomes harder as `0x00` (end-of-file) clashes with `0x20` (space) and `0x09` (tab) clashes with `)` (ends-grouped). While it's not necessary to classify which character is found, we need more operations to do so. The minimum amount I could come up with:

```
__m128i input = _mm_loadu_si128(pointer);
__m128i nibbles = _mm_and_si128(input, _mm_set1_epi8(0x0f));
__m128i compare = _mm_shuffle_epi8(_mm_load_si128(contiguous_table), nibbles);
// contiguous_and is a shuffle table that disables 0x20 for characters that clash
__m128i and = _mm_shuffle_epi8(_mm_load_si128(contiguous_and), nibbles);
input = _mm_and_si128(input, and);
__m128i result = _mm_cmpeq_epi8(input, compare);
```

Gut feeling, this is too much operations to outweigh the second index, but I'm happy to be proven wrong :slightly_smiling_face:

Now, it's not strictly necessary to try and locate everything as all structural characters will have indexes (note that this is not the case for `;`, or comments). Therefore, it may be feasible to not write the terminating character for fields and use the start of the next field as a maximum. That way, we only need to search for the `;` and space characters, which do not clash and therefore allows for a single `_mm_shuffle_epi8` followed by a `_mm_cmpeq_epi8`.

This would allow us to drop the second index, which may, or may not outweigh the extra writes.

For names and strings this would work just fine. For numbers, like ttl values, we can simply check if the finishing character is considered contiguous or not. For record-types, this makes things slightly more complicated as the currently "perfect-hash" algorithm actually needs the length (see [scan_type](https://github.com/NLnetLabs/simdzone/blob/81e0902a39d7dbfedb2a331f858abf5fbe294b66/src/parser.h#L194-L242)). The same is true for integer fields that allow for symbolic values, e.g. the [Hash Algorithm in NSEC3](https://www.rfc-editor.org/rfc/rfc5155.html#section-3.1.1) (see [parse_int8](https://github.com/NLnetLabs/simdzone/blob/81e0902a39d7dbfedb2a331f858abf5fbe294b66/src/parser.h#L378-L403)).

It's likely scanning for svc_param fields specified in [draft-ietf-dnsop-svcb-https-11](https://datatracker.ietf.org/doc/draft-ietf-dnsop-svcb-https/), soon to be an RFC will be more complicated too. Reason being that the scanner will see `key="value"` as two separate tokens that need be combined in a specialized `lex` function. The algorithm is currently pretty straightforward, i.e. check if the closing index is the start of a quoted section, as no whitespace is allowed between the key and value. That algorithm will become somewhat more complex once there's no more length index.

Long story short, it's possible to drop the second index, but it may not be worth the effort. Either we need more than one SIMD operation to locate the end of a contiguous field or require a second load in `lex` to find an absolute maximum.

**Comment by k0ekk0ek** [comment]
Turns out that the number of operations in the scanner with the length index and keeping track of embedded newlines is between 40%-50% slower than what I started out with and that is without writing the token code, which is needed to deny quoted fields where it they do not make sense. e.g. TYPE, TTL. That is also without fancier tracking of terminating index. Currently the last index is discarded if the token is not yet complete, but it may be that the last two indexes must be discarded. e.g. for `foo"bar\0`, which will be more common with SVCB. I did a really quick [prototype](https://github.com/k0ekk0ek/scanner) (eventually merged, then removed) that does not suffer from the same problems and does get me ~2GB/s. So switching to a single index is almost certainly faster and tackles some problems that are otherwise iffy to solve.
